mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/AttachmentLoader.java

```diff
@@ -46,21 +47,24 @@ public AttachmentLoader(CassandraAttachmentMapper attachmentMapper) {
         this.attachmentMapper = attachmentMapper;
     }
 
-    public CompletableFuture<Stream<SimpleMailboxMessage>> toMailboxMessageWithAttachments(
-                CompletableFuture<Stream<Pair<MessageWithoutAttachment, Stream<MessageAttachmentRepresentation>>>> messageRepresentations) {
+    public CompletableFuture<Stream<SimpleMailboxMessage>> addAttachmentToMessages(Stream<Pair<MessageWithoutAttachment,
+            Stream<MessageAttachmentRepresentation>>> messageRepresentations, MessageMapper.FetchType fetchType) {
 
-        return FluentFutureStream.of(messageRepresentations)
-            .thenComposeOnAll(pair -> getAttachments(pair.getRight().collect(Guavate.toImmutableList()))
-                    .thenApply(attachments -> Pair.of(pair.getLeft(), attachments)))
-            .map(pair ->
-                pair.getLeft()
-                    .toMailboxMessage(pair.getRight()
-                        .stream()
-                        .collect(Guavate.toImmutableList())))
-            .completableFuture();
+        if (fetchType == MessageMapper.FetchType.Body || fetchType == MessageMapper.FetchType.Full) {
+            return FluentFutureStream.of(
+                messageRepresentations
+                    .map(pair -> getAttachments(pair.getRight().collect(Guavate.toImmutableList()))
+                        .thenApply(attachments -> pair.getLeft().toMailboxMessage(attachments))))
+                .completableFuture();
+        } else {
+            return CompletableFuture.completedFuture(messageRepresentations
+                .map(pair -> pair.getLeft()
+                    .toMailboxMessage(ImmutableList.of())));
+        }
     }
 
-    public CompletableFuture<Collection<MessageAttachment>> getAttachments(List<MessageAttachmentRepresentation> attachmentRepresentations) {
+    @VisibleForTesting
+    CompletableFuture<List<MessageAttachment>> getAttachments(List<MessageAttachmentRepresentation> attachmentRepresentations) {
```

Comment:
You should directly take a stream here IMO.
Because in the current state when you call it you convert a stream to a list (line 56) before calling it and inside you revert this list to a stream!